cpp/src/arrow/json/converter.cc

```diff
@@ -123,13 +123,11 @@ class NumericConverter : public PrimitiveConverter {
     RETURN_NOT_OK(builder.Resize(dict_array.indices()->length()));
 
     auto visit_valid = [&](string_view repr) {
-      value_type value;
-      if (!internal::ParseValue(numeric_type_, repr.data(), repr.size(), &value)) {
-        return GenericConversionError(*out_type_, ", couldn't parse:", repr);
+      if (auto value = internal::ParseValue(numeric_type_, repr)) {
```

Comment:
This is nice that you fixed this "maybe-undefined" `value`. There are two more like this in `csv/converter.cc`: can you fix those too? cf. https://github.com/r-windows/rtools-packages/runs/1277507717#step:4:893

(How difficult would it be to add a CI job that fails on `-Wmaybe-undefined`? I feel like every time I inspect the C++ build logs on the R mingw builds (which is infrequent), I find more warnings like this.

Reply:
I think these can be fixed, but I'd like to wait for a follow up so we don't conflate any perf impact of this PR with unrelated changes inside CSV